Performance degraded with yavijava 6.0.

Hi,

We manage dvSwitch by taking its configuration backup and deploying changes.
We do configuration backup by running lot of queries. With vijava the whole operation of backup was taking 6 seconds and when we change to yavijava 6.0.01 jar , the same operation took 9 minutes.
I will try to summarize what exactly we do to take configuration backup of DVSwitch.

``` java
DistributedVirtualSwitch dvSwitch = state.dvSwitch;
DVSConfigInfo dvsCfg = dvSwitch.getConfig();
int[] vlanIds = dvSwitch.queryUsedVlanIdInDvs();
ManagedObjectReference[] uplinks = dvsCfg.getUplinkPortgroup();
String description = dvsCfg.getDescription();
String ipAddress = dvsCfg.getSwitchIpAddress();
 dvsCfg.getNumPorts()
 + dvsCfg.getMaxPorts()
+ dvsCfg.getNumStandalonePorts()
```

The above given are some methods which we trigger to get confguration of dvSwitch. We trigger many more such methods to get complete information of all dvPort groups and all ports associated with port-groups.

So if I use yavijava 6.0.01 jar it works in dvSwitch but with degraded performance. When I try to run same operation on dvSwitch which is on vSphere 5.5 it takes 9 minutes as compared to 6 seconds in case of vijava.

Kindly suggest.

Ok I try to build with steps provided by you. The build is failing.

oEx.java:55: warning - @return tag has no arguments.
34 warnings
:javadocJar
:sourcesJar
:signArchives FAILED

FAILURE: Build failed with an exception.
- What went wrong:
  Execution failed for task ':signArchives'.
  
  > Cannot perform signing task ':signArchives' because it has no configured signa
  > tory
- Try:
  Run with --stacktrace option to get the stack trace. Run with --info or --debug
  option to get more log output.

BUILD FAILED

Total time: 4 mins 16.203 secs

L:\Git\yavijava>gradlew build --stacktrace
:compileJava UP-TO-DATE
:compileGroovy UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar UP-TO-DATE
:javadoc UP-TO-DATE
:javadocJar UP-TO-DATE
:sourcesJar UP-TO-DATE
:signArchives FAILED

FAILURE: Build failed with an exception.
- What went wrong:
  Execution failed for task ':signArchives'.
  
  > Cannot perform signing task ':signArchives' because it has no configured signa
  > tory
- Try:
  Run with --info or --debug option to get more log output.
- Exception is:
  org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':signArc
  hives'.
      at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.ex
  ecuteActions(ExecuteActionsTaskExecuter.java:69)
      at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.ex
  ecute(ExecuteActionsTaskExecuter.java:46)
      at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExec
  uter.execute(PostExecutionAnalysisTaskExecuter.java:35)
      at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.exec
  ute(SkipUpToDateTaskExecuter.java:64)
      at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execut
  e(ValidatingTaskExecuter.java:58)
      at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecu
  ter.execute(SkipEmptySourceFilesTaskExecuter.java:42)
      at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter
  .execute(SkipTaskWithNoActionsExecuter.java:52)
      at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execut
  e(SkipOnlyIfTaskExecuter.java:53)
      at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter
  .execute(ExecuteAtMostOnceTaskExecuter.java:43)
      at org.gradle.api.internal.AbstractTask.executeWithoutThrowingTaskFailur
  e(AbstractTask.java:296)
      at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorW
  orker.executeTask(AbstractTaskPlanExecutor.java:79)
      at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorW
  orker.processTask(AbstractTaskPlanExecutor.java:63)
      at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorW
  orker.run(AbstractTaskPlanExecutor.java:51)
      at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(Defaul
  tTaskPlanExecutor.java:23)
      at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(Defau
  ltTaskGraphExecuter.java:86)
      at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTask
  ExecutionAction.java:29)
      at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecute
  r.java:61)
      at org.gradle.execution.DefaultBuildExecuter.access$200(DefaultBuildExec
  uter.java:23)
      at org.gradle.execution.DefaultBuildExecuter$2.proceed(DefaultBuildExecu
  ter.java:67)
      at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildEx
  ecutionAction.java:32)
      at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecute
  r.java:61)
      at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecute
  r.java:54)
      at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(Default
  GradleLauncher.java:148)
      at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradle
  Launcher.java:105)
      at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLaun
  cher.java:85)
      at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildCon
  troller.run(InProcessBuildActionExecuter.java:81)
      at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.jav
  a:33)
      at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.jav
  a:24)
      at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProce
  ssBuildActionExecuter.java:39)
      at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProce
  ssBuildActionExecuter.java:29)
      at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:50)
      at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.jav
  a:171)
      at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.
  execute(CommandLineActionFactory.java:237)
      at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.
  execute(CommandLineActionFactory.java:210)
      at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRunti
  meValidationAction.java:35)
      at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRunti
  meValidationAction.java:24)
      at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(
  CommandLineActionFactory.java:206)
      at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(
  CommandLineActionFactory.java:169)
      at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionRep
  ortingAction.java:33)
      at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionRep
  ortingAction.java:22)
      at org.gradle.launcher.Main.doAction(Main.java:33)
      at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
      at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBoots
  trap.java:54)
      at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.j
  ava:35)
      at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
      at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.ja
  va:30)
      at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:127)
      at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:56)
  Caused by: org.gradle.api.InvalidUserDataException: Cannot perform signing task
  ':signArchives' because it has no configured signatory
      at org.gradle.plugins.signing.Sign.generate(Sign.groovy:162)
      at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:63)
      at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskF
  actory$StandardTaskAction.doExecute(AnnotationProcessingTaskFactory.java:218)
      at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskF
  actory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:211)
      at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskF
  actory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:200)
      at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(Abstra
  ctTask.java:570)
      at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(Abstra
  ctTask.java:553)
      at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.ex
  ecuteAction(ExecuteActionsTaskExecuter.java:80)
      at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.ex
  ecuteActions(ExecuteActionsTaskExecuter.java:61)
      ... 47 more

--- a/build.gradle
+++ b/build.gradle
@@ -8,7 +8,8 @@ apply plugin: 'groovy'
 
 group = "com.toastcoders"
 archivesBaseName = "yavijava"
-version = '6.0.01'
+version = '6.0.01-SNAPSHOT'
+ext.isReleaseVersion = !version.endsWith("SNAPSHOT")
 targetCompatibility = 1.6
 sourceCompatibility = 1.6
 
@@ -86,6 +87,7 @@ artifacts {
 }
 
 signing {
+    required { isReleaseVersion && gradle.taskGraph.hasTask("uploadArchives") }
     sign configurations.archives
 }
 

--- a/src/intTest/java/com/vmware/vim25/ws/WSClientIntTest.java
+++ b/src/intTest/java/com/vmware/vim25/ws/WSClientIntTest.java
@@ -1,22 +1,29 @@
 package com.vmware.vim25.ws;
 
+import java.io.IOException;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.rmi.RemoteException;
 import java.util.Calendar;
-import javax.net.ssl.SSLHandshakeException;
 
-import com.vmware.vim25.*;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.SSLSocketFactory;
+
 import org.apache.log4j.Logger;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
 import com.utility.LoadVcenterProps;
+import com.vmware.vim25.ManagedObjectReference;
+import com.vmware.vim25.ObjectContent;
+import com.vmware.vim25.ObjectSpec;
+import com.vmware.vim25.PropertyFilterSpec;
+import com.vmware.vim25.PropertySpec;
+import com.vmware.vim25.SelectionSpec;
 import com.vmware.vim25.mo.ServiceInstance;
 import com.vmware.vim25.mo.util.PropertyCollectorUtil;
-
-import javax.net.ssl.HttpsURLConnection;
 
 public class WSClientIntTest {
 
@@ -46,7 +53,6 @@ public class WSClientIntTest {
         } catch (MalformedURLException e) {
             e.printStackTrace();
         }
-
         if(si != null) {
             wsClient = new WSClient(LoadVcenterProps.url, true);
 
@@ -140,6 +146,29 @@ public class WSClientIntTest {
     }
 
     /**
+     * This test will confirm that the internal SSL socket factory is initiate only once in the WSClient (Issue #38).
+     */
+    @Test
+    public void testSSLSocketFactoryInitialization() throws Exception {
+        CustomWSClient client = new CustomWSClient(LoadVcenterProps.url, true);
+        Assert.assertEquals(1, createdSSLFactory);
+
+        try {
+            client.invoke("RetrieveProperties", buildGetHostsArgs(), "ObjectContent[]");
+        } catch (RemoteException e) {
+        }
+        
+        Assert.assertEquals(1, createdSSLFactory);
+
+        try {
+            client.invoke("RetrieveProperties", buildGetHostsArgs(), "ObjectContent[]");
+        } catch (RemoteException e) {
+        }
+        
+        Assert.assertEquals(1, createdSSLFactory);
+    }
+
+   /**
      * This method will build the request payload.
      * 
      * @return Argument[]
@@ -178,5 +207,31 @@ public class WSClientIntTest {
                 new PropertyFilterSpec[] { spec });
 
         return paras;
+    }
+
+    /**
+     * Counter for created factory in {@link CustomWSClient}.
+     */
+    private int createdSSLFactory = 0;
+    
+    /**
+     * This extension of the WSClient will create count the number of time the {@link SSLSocketFactory} was created.
+     * 
+     * @author Francis Beaulé
+     *
+     */
+    private class CustomWSClient extends WSClient {
+        public CustomWSClient(String serverUrl, boolean ignoreCert) throws MalformedURLException, IOException {
+            super(serverUrl, ignoreCert);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected SSLSocketFactory getSocketFactory(boolean ignoreCert) throws IOException {
+           ++createdSSLFactory;
+            return super.getSocketFactory(ignoreCert);
+        }
     }
 }

--- a/src/main/java/com/vmware/vim25/ws/WSClient.java
+++ b/src/main/java/com/vmware/vim25/ws/WSClient.java
@@ -34,6 +34,7 @@ package com.vmware.vim25.ws;
 import org.apache.log4j.Logger;
 
 import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLSocketFactory;
 import java.io.*;
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
@@ -52,22 +53,22 @@ import java.text.MessageFormat;
 public class WSClient extends SoapClient {
 
     private static final Logger log = Logger.getLogger(WSClient.class);
-    private boolean ignoreCert = false;
+    private final SSLSocketFactory sslSocketFactory;
 
     XmlGen xmlGen = new XmlGenDom();
 
-    public WSClient(String serverUrl) throws MalformedURLException {
+    public WSClient(String serverUrl) throws MalformedURLException, RemoteException {
         this(serverUrl, true);
     }
 
-    public WSClient(String serverUrl, boolean ignoreCert) throws MalformedURLException {
+    public WSClient(String serverUrl, boolean ignoreCert) throws MalformedURLException, RemoteException {
         if (serverUrl.endsWith("/")) {
             serverUrl = serverUrl.substring(0, serverUrl.length() - 1);
         }
         log.trace("Creating WSClient to server URL: " + serverUrl);
         log.trace("Ignore ssl: " + ignoreCert);
         this.baseUrl = new URL(serverUrl);
-        this.ignoreCert = ignoreCert;
+        this.sslSocketFactory = getSocketFactory(ignoreCert);
     }
 
     public Object invoke(String methodName, Argument[] paras, String returnType) throws RemoteException {
@@ -115,22 +116,9 @@ public class WSClient extends SoapClient {
     }
 
     protected InputStream post(String soapMsg) throws IOException {
-        HttpURLConnection postCon;
-
-        if(ignoreCert && trustManager != null) {
-            log.warn("The option to ignore certs has been set along with a provided trust manager. This is not a valid scenario and the trust manager will be ignored.");
-        }
-
-        if (baseUrl.getProtocol().equalsIgnoreCase("https") && ignoreCert) {
-            postCon = (HttpsURLConnection) baseUrl.openConnection();
-            ((HttpsURLConnection) postCon).setSSLSocketFactory(TrustAllSSL.getTrustContext().getSocketFactory());
-        } else if(baseUrl.getProtocol().equalsIgnoreCase("https") && !ignoreCert) {
-            postCon = (HttpsURLConnection) baseUrl.openConnection();
-            if(trustManager != null) {
-                ((HttpsURLConnection) postCon).setSSLSocketFactory(CustomSSLTrustContextCreator.getTrustContext(trustManager).getSocketFactory());
-            }
-        } else {
-            postCon = (HttpURLConnection) baseUrl.openConnection();
+        HttpURLConnection postCon = (HttpURLConnection) baseUrl.openConnection();
+        if (sslSocketFactory != null && baseUrl.getProtocol().equalsIgnoreCase("https")) {
+            ((HttpsURLConnection) postCon).setSSLSocketFactory(sslSocketFactory);
         }
 
         log.trace("POST: " + soapAction);
@@ -206,4 +194,7 @@ public class WSClient extends SoapClient {
         return new OutputStreamWriter(os, "UTF8");
     }
 
+    protected SSLSocketFactory getSocketFactory(boolean ignoreCert) throws RemoteException {
+        return ignoreCert ? TrustAllSSL.getTrustContext().getSocketFactory() : null;
+    }
 }

